### DISCUSSION TITLE
XML TREE TEASER

### DISCUSSION THREAD

**Discussion by gdiana** [grafana/grafana-infinity-datasource#303] [General]
Here is the XML teaser you asked me to post. Difficult to get XML and UQL to work. works great in JSONATA..
[XMLSAMPLE.txt](https://github.com/yesoreyeram/grafana-infinity-datasource/files/8542187/XMLSAMPLE.txt)

**Comment by yesoreyeram** [comment]
So what data you are interested in the above json? (Or JSONata query you executed)

BTW, have you tried v0.8.8 which have JSONata support?

```
parse-json
| jsonata "xml.meters.meter[cid!=175]"
```

<img width="1212" alt="image" src="https://user-images.githubusercontent.com/153843/164743496-d25d3f2d-2c93-464e-8aeb-123806149324.png">

**Reply by gdiana** [reply]
No I have not tried your JASONATA yet only saw it yesterday but will try after I upgrade. JASONATA is very powerful indeed.

I attached the XML above as a file as I see it missed a whole lot.

The above XML works fine using XML and the path xml.meters[0].meter[0].dailyreads[0]
then I can get date etc etc for chart. But p1 which is all the days data is a single string and needs to be converted to an array. It assumes meters are read every 30 min or 48 samples per day so just provides the 30 min data and the user must do the rest. I do not expect you to try and solve but just show some of the curveballs that APIs throw at us.

But when doing the same in UQL the same path (| scope xml.meters.meter.dailyreads) does not work for some reason.

If I try and go further down  xml.meters[0].meter[0].dailyreads[0].vidata[0].timeitme neither XML or UQL works but extrernal JSONATA works.

Some other queries.

Maybe an XML to JSON converter might be worth considering?

Please provide a short example/s on how the  parse_url() function is supposed to work and be used? 

I thought this would be a way to get the path to an individual column or item of data but maybe I am wrong.

Thanks.

**Reply by yesoreyeram** [reply]
yea.. splitting a string is not supported in UQL yet. But looks like an interesting usecase. Should easy to support in next few versions of UQL.  

BTW, how do you achieve this with JSONata?

**Reply by yesoreyeram** [reply]
XML to JSON convertor? that's what the `parse-xml` command does. ( converts the xml into json format then so you can apply UQL commands on top of it )

`parse_url` function used to extract part of a URL string. say you want to extract domain name from a URL string/ specific query parameter of the URL, then you can use this method.

**Comment by yesoreyeram** [comment]
@gdiana  - [here](https://grafana-infinity-datasource.herokuapp.com/explore?orgId=1&left=%7B%22datasource%22:%22Infinity%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22type%22:%22uql%22,%22source%22:%22url%22,%22format%22:%22timeseries%22,%22url%22:%22https:%2F%2Fgist.githubusercontent.com%2Fyesoreyeram%2Fa008ec91667e7bb5a4ebb0dc506a5a47%2Fraw%2F88fc510f78e1b23f58a14da3a57d28239ccb867d%2Fxml-teaser.json%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22uql%22:%22parse-json%5Cn%7C%20scope%20%5C%22xml.meters.meter%5C%22%5Cn%7C%20extend%20%5C%22timeitems%5C%22%3D%5C%22dailyreads.dailyread.vidata.timeitem%5C%22%5Cn%7C%20mv-expand%20%5C%22timeItem%5C%22%3D%5C%22timeitems%5C%22%5Cn%7C%20extend%20%5C%22time%5C%22%3Dtodatetime(%5C%22timeItem.time%5C%22),%20%5C%22red_i%5C%22%3D%5C%22timeItem.red.i%5C%22%5Cn%7C%20project%20%5C%22time%5C%22,%20strcat(%5C%22cid%5C%22,%27--%27,%5C%22serial%5C%22),%20%5C%22red_i%5C%22%22%7D%5D,%22range%22:%7B%22from%22:%221650411084407%22,%22to%22:%221650460438038%22%7D%7D) is how i do graph the data from the json you provided

```kusto
parse-json
| scope "xml.meters.meter"
| extend "timeitems"="dailyreads.dailyread.vidata.timeitem"
| project-away "dailyreads"
| mv-expand "timeItem"="timeitems"
| extend "time"=todatetime("timeItem.time"), "red_i"="timeItem.red.i"
| project "time", "red_i", strcat("cid",'--',"serial")
```
<img width="1613" alt="image" src="https://user-images.githubusercontent.com/153843/164745240-70f0d1b5-af24-4025-b486-44c853a7a143.png">

**Reply by gdiana** [reply]
Ooh this is cool, thanks. I will play around.

Sometimes, one has to get ones head around the query logic and syntax and I am no UQL expert.

**Reply by gdiana** [reply]
Hey I notice you are using parse-json and not parse-xml?

Should that work? It seems to though?

**Reply by yesoreyeram** [reply]
no worries @gdiana . UQL is a custom language I built specifically for this purpose. So I can't expect anyone to know this better. Glad at least you are able to use at some extent.

**Reply by yesoreyeram** [reply]
As far as I know, apart from me there are only couple of people can write UQL ( including you ) 😄

**Reply by yesoreyeram** [reply]
parse-xml should work. ( i use parse-json becuase you shared json earlier )

**Reply by yesoreyeram** [reply]
```kusto
parse-xml
| scope "meters.meter"
| extend "timeitems"="dailyreads.dailyread"
| mv-expand "timeItem"="timeitems"
| extend "time"=todatetime("timeItem.date"), "kva"="timeItem.kva"
| project "time", strcat("cid",'--',"serial"), "kva"
```
<img width="1531" alt="image" src="https://user-images.githubusercontent.com/153843/164750966-75a620e4-08a5-4ebd-b2da-163adfe20443.png">

**Reply by gdiana** [reply]
Here is how I am currently using it for other API calls.

![image](https://user-images.githubusercontent.com/13445152/164753840-7e8a9619-7965-443f-ba07-b45f08ee418e.png)

**Reply by gdiana** [reply]
Here is another where I use variables to parse and use the format_datetime() function to provide me energy usage per day of month

I have found the format_datetime() is limited. For example week of year is not yet available.
I am using the format(0) function to bypass the nested by. Maybe this can be done but I am not aware so I just found a workaround.

![image](https://user-images.githubusercontent.com/13445152/164754356-611510c4-285e-496b-af60-c83e7d058105.png)

**Reply by yesoreyeram** [reply]
I can approve you as a power user of UQL 😄 

BTW, can you create a feature request for `week_of_the_year` macro in https://github.com/yesoreyeram/uql ? and also file any relevant bugs/feature requests there.

**Reply by gdiana** [reply]
And here is another where I found the average energy usage per day for a month or time period involved. Again used format_datetime() function.

![image](https://user-images.githubusercontent.com/13445152/164756155-cc62a1b7-4dbc-4190-b471-581ba3d446dd.png)


there are problems as UQL can do orders by day number (1,2,3,4 ...) but not by (Mon, Tues, Wed etc) 

So I had to force use of numbers to get order by to work.

In your example above you can see the hour per day are not correctly ordered and it should start from 00h00 ....... 23h30

IN SQL one may use the GROUP function ie GROUP day, month to do this, but I am not sure with UQL and whether a group function even exists, but for now I am getting it to work 👍

**Reply by yesoreyeram** [reply]
In UQL group by function exist. (summarize is the group by function in UQL ) and you may need to use `startofweek` function before summarize to get the results group by week.

**Reply by yesoreyeram** [reply]
one more hint. UQL uses https://day.js.org/docs/en/display/format for formatting, So to format a date, you can use any of the suggested options.

**Reply by gdiana** [reply]
In the above, a common thing is to try and do a calendar chart, where repeat in the X-AXIS is DOW and repeat in the Y-AXIS is Week of year.

This is very easy to do using Tableau but would require some doing in GRAFANA. It would require parsing a UQL variable from the query to become a repeat value but is probably a huge misison.

**Reply by gdiana** [reply]
> one more hint. UQL uses https://day.js.org/docs/en/display/format for formatting, So to format a date, you can use any of the suggested options.

I see the things I require need the advanced format which require some PLUGINS. I tried a few but it just returns the string I enter so appears the PLUGINS are required

![image](https://user-images.githubusercontent.com/13445152/164777062-1ef90be5-a7a7-458d-bd62-9868a144eea9.png)

**Reply by gdiana** [reply]
> @gdiana - [here](https://grafana-infinity-datasource.herokuapp.com/explore?orgId=1&left=%7B%22datasource%22:%22Infinity%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22type%22:%22uql%22,%22source%22:%22url%22,%22format%22:%22timeseries%22,%22url%22:%22https:%2F%2Fgist.githubusercontent.com%2Fyesoreyeram%2Fa008ec91667e7bb5a4ebb0dc506a5a47%2Fraw%2F88fc510f78e1b23f58a14da3a57d28239ccb867d%2Fxml-teaser.json%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22uql%22:%22parse-json%5Cn%7C%20scope%20%5C%22xml.meters.meter%5C%22%5Cn%7C%20extend%20%5C%22timeitems%5C%22%3D%5C%22dailyreads.dailyread.vidata.timeitem%5C%22%5Cn%7C%20mv-expand%20%5C%22timeItem%5C%22%3D%5C%22timeitems%5C%22%5Cn%7C%20extend%20%5C%22time%5C%22%3Dtodatetime(%5C%22timeItem.time%5C%22),%20%5C%22red_i%5C%22%3D%5C%22timeItem.red.i%5C%22%5Cn%7C%20project%20%5C%22time%5C%22,%20strcat(%5C%22cid%5C%22,%27--%27,%5C%22serial%5C%22),%20%5C%22red_i%5C%22%22%7D%5D,%22range%22:%7B%22from%22:%221650411084407%22,%22to%22:%221650460438038%22%7D%7D) is how i do graph the data from the json you provided
> 
> ```kusto
> parse-json
> | scope "xml.meters.meter"
> | extend "timeitems"="dailyreads.dailyread.vidata.timeitem"
> | project-away "dailyreads"
> | mv-expand "timeItem"="timeitems"
> | extend "time"=todatetime("timeItem.time"), "red_i"="timeItem.red.i"
> | project "time", "red_i", strcat("cid",'--',"serial")
> ```
> 
> <img alt="image" width="1613" src="https://user-images.githubusercontent.com/153843/164745240-70f0d1b5-af24-4025-b486-44c853a7a143.png">

Hi,

I have tried and tried and tried but I am unable to replicate this. The first extend does not work and return anything.

What you have done is allow access at different levels which is great but I just get {} for timeitems.

I understand you are busy but I have not found it at all possible to replicate. I am using the latest version (1.0)
